x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -187,6 +196,31 @@ public void clusterChanged(ClusterChangedEvent event) {
         }
     }
 
+    private void logMlNodeHeterogeneity() {
+        ActionListener<Set<String>> architecturesListener = new ActionListener<Set<String>>() {
+            @Override
+            public void onResponse(Set<String> architectures) {
+                if (architectures.size() > 1) {
+                    String architecturesList = String.join(", ", architectures);
+                    logger.warn(
+                        format(
+                            "Heterogeneous platform architectures were detected among ML nodes. "
+                                + "This will prevent the deployment of some trained models. Distinct platform architectures detected: [%s]",
+                            architecturesList
+                        )
+                    );
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error("Failed to detect heterogeneity among ML nodes with exception: ", e);
+            }
+        };
+
+        MLPlatformArchitecturesUtil.getNodesOsArchitectures(threadPool, client, architecturesListener);
+    }
```

Comment:
Do we have a standard way of testing for log messages? IIRC, we need some special set up for testing these since it uses static methods.

Reply:
There's an example of how to do it in `IndexingSlowLogTests`: https://github.com/elastic/elasticsearch/blob/a830787b07be39d53d51623fdf6ad8b7dde54ec5/server/src/test/java/org/elasticsearch/index/IndexingSlowLogTests.java#L54-L68

You temporarily add another appender to the logging framework for the duration of your tests, and then you can assert on the last log message that appender was told to log.